elasticdl/docker/Dockerfile

```diff
@@ -1,18 +1,108 @@
-FROM ${BASE_IMAGE}
+ARG EXTRA_PYPI_INDEX=https://pypi.org/simple
+
+# Stage 1: builder
+FROM ${BASE_IMAGE} as builder
+
+RUN apt update && \
+    apt install -y unzip curl git software-properties-common g++ wget \
+    libeigen3-dev
+
+COPY elasticdl/requirements.txt /requirements.txt
+RUN pip install -r /requirements.txt --extra-index-url=${EXTRA_PYPI_INDEX} && rm /requirements.txt
+
+ENV TF_PATH /tmp/tensorflow
+RUN cd /tmp && git clone --depth=1 https://github.com/tensorflow/tensorflow.git
+
+# Install Go and related tools
+ARG GO_MIRROR_URL=https://dl.google.com/go
+ENV GOPATH /root/go
+ENV PATH /usr/local/go/bin:$GOPATH/bin:$PATH
+COPY elasticdl/docker/scripts/install-go.bash /
+RUN /install-go.bash ${GO_MIRROR_URL} && rm /install-go.bash
+
+# Install protobuf and protoc
+COPY elasticdl/docker/scripts/install-protobuf.bash /
+RUN /install-protobuf.bash && rm /install-protobuf.bash
+
+# Install elasticdl.org/elasticdl Go package
+ENV ELASTICDLPATH /root/elasticdl
+COPY . /elasticdl
+RUN cd /elasticdl && \
+    make -f elasticdl/Makefile && \
+    cp -r /elasticdl/elasticdl/pkg $ELASTICDLPATH && \
+    cd $GOPATH/pkg/mod/github.com/tensorflow && \
+    go mod init github.com/tensorflow &&\
+    cd $ELASTICDLPATH && \
+    go mod init elasticdl.org/elasticdl && \
+    go mod edit -replace github.com/tensorflow=${GOPATH}/pkg/mod/github.com/tensorflow && \
+    go mod tidy && \
+    go install ./... 
+
+# Create elasticdl package
+RUN cd /elasticdl && \
+    python setup.py bdist_wheel && \
+    cp dist/elasticdl-develop-py3-none-any.whl / 
+
+# Stage 2: data
+FROM builder as data
+
+COPY elasticdl/requirements-dev.txt /requirements-dev.txt
+RUN pip install -r /requirements-dev.txt --extra-index-url=${EXTRA_PYPI_INDEX} && rm /requirements-dev.txt
+
+# This assumes that the data generation package is independent with the
+# rest part of ElasticDL.  The generated data will be in /data.
+COPY elasticdl/python/data/recordio_gen/image_label.py /var/image_label.py
+RUN python /var/image_label.py --dataset mnist --fraction 0.15 \
+        --records_per_shard 4096 /data
+
+# Copy frappe dataset
+COPY elasticdl/python/data/recordio_gen/frappe_recordio_gen.py /var/frappe_recordio_gen.py
+RUN python /var/frappe_recordio_gen.py --data /root/.keras/datasets --output_dir /data/frappe \
+    --fraction 0.05
+# Copy heart dataset
+COPY elasticdl/python/data/recordio_gen/heart_recordio_gen.py /var/heart_recordio_gen.py
+RUN python /var/heart_recordio_gen.py --data_dir /root/.keras/datasets --output_dir /data/heart
+# Copy census dataset
+COPY elasticdl/python/data/recordio_gen/census_recordio_gen.py /var/census_recordio_gen.py
+RUN python /var/census_recordio_gen.py --data_dir /root/.keras/datasets --output_dir /data/census
+
+RUN rm -rf /root/.keras/datasets
+
+# Stage 3: dev
+FROM data as dev
+
+# Remove /elasticdl in dev
+RUN cd / && rm -rf /elasticdl && \
+    rm /elasticdl-develop-py3-none-any.whl
+
+# Stage 4: ci
```

Comment:
It seems that dev and ci are very thin wrapper. Could we just has two stage, dev and release?
- The dev install necessary dependencies, and generate Go binary and python wheel package.
- The release just install the Go binary and python wheel package, and copy model zoo.

Reply:
ci is used in ci testing in [client_test.sh]( https://github.com/sql-machine-learning/elasticdl/blob/develop/scripts/client_test.sh). It needs the source code for image building.

So we need at least 3 images:
1. dev that including all building/development dependencies. Also some sample data(maybe move sample data to ci).
2. ci to including both dev and source code and sample model zoo.
3. release for the end-user that does not need those dependencies and sample data. Does it need sample model zoo?

If dev does not include sample data, we can reduce to 3 stages.
1. dev with Go binary and python wheel.
2. release to install binary/wheel from dev.
3. ci, based on dev, add sample model zoo and sample data.